builders/build/substrate/adding-custom-made-module.md

```diff
@@ -0,0 +1,347 @@
+
+The `#[pallet::error]` macro is used to annotate an enumeration of potential errors that could occur during execution. It is crucial for security to ensure that all error situations are handled gracefully, without causing the runtime to crash.
+
+The following example of this macro implementation shows the errors that might occur in the lottery module:
+
+```rust
+// Errors inform users that something went wrong.
+#[pallet::error]
+pub enum Error<T> {
+    NotEnoughCurrency,
+    AccountAlreadyParticipating,
+    CanNotAddParticipant,
+}
+```
+
+### Implementing Events {: #implementing-events }
+
+The `#[pallet::event]` macro is applied to an enumeration of events to inform the user of any changes in the state or important actions that happened during the execution in the runtime.
+
+As an example, for the `lottery-example` module, this macro could be configured with the following events:
+
+```rust
+#[pallet::event]
+#[pallet::generate_deposit(pub(super) fn deposit_event)]
+pub enum Event<T: Config> {
+    /// Event emitted when a ticket is bought
+    TicketBought { who: T::AccountId },
+    /// Event emitted when the prize is awarded
+    PrizeAwarded { winner: T::AccountId },
+    /// Event emitted when there are no participants
+    ThereAreNoParticipants,
+}
+```
+
+### Implementing Storage for State Persistence {: #implementing-storage }
+
+The `#[pallet::storage]` macro initializes a runtime storage structure.  In the heavily constrained environment of an AppChain, deciding what to store and which structure to use can be critical in terms of performance. More on this topic is covered in the [Substrate documentation](https://docs.substrate.io/build/runtime-storage/){target=_blank}.
+
+In this example,  the `lottery-example` module needs a basic value storage structure used to persist the list of participants in a bounded capacity vector ([BoundedVec](https://crates.parity.io/frame_support/storage/bounded_vec/struct.BoundedVec.html){target=_blank}). This can be initialized as follows:
+
+```rust
+#[pallet::storage]
+#[pallet::getter(fn get_participants)]
+pub(super) type Participants<T: Config> = StorageValue<
+    _,
+    BoundedVec<T::AccountId, T::MaxParticipants>,
+    OptionQuery
+>;
+```
+### The Complete Module {: #complete-module }
+
+Putting all the pieces together, after implementing all the required macros and adding the custom logic, now the module is complete and ready to be used in the runtime. 
+
+??? code "View the complete module file"
+
+    ```rust
+    --8<-- 'code/modules/lottery-example.rs'
+    ```
+
+## Configure the Runtime {: #configure-runtime }
+
+Finally, with the module finished, it can be included in the runtime. By doing so, the transactions `buy_tickets` and `award_prize` will be callable by the users. This also means that the (Polkadot.js API)(/builders/interact/substrate-api/polkadot-js-api/) will be decorated with this module and all the available calls that it contains.
```

Comment:
```suggestion
Finally, with the module finished, it can be included in the runtime. By doing so, the transactions `buy_tickets` and `award_prize` will be callable by the users. This also means that the [Polkadot.js API](/builders/interact/substrate-api/polkadot-js-api/){target=_blank} will be decorated with this module and all the available calls that it contains.
```